### PR TITLE
wallet: add a configure function

### PR DESCRIPTION
**Pull request by moneromooo-monero**
It goes through configuration parameters and calls a function,
describing the parameter.

### CODE REVIEW

File: src/simplewallet/simplewallet.cpp [hunk 1]
```diff
@@ -245,17 +245,18 @@ namespace
...
+  std::string input_line(const std::string& prompt, const std::vector<std::string> &values = {})
   {
 #ifdef HAVE_READLINE
     rdln::suspend_readline pause_readline;
 #endif
     std::cout << prompt;
-    if (yesno)
-      std::cout << "  (Y/Yes/N/No)";
+    if (!values.empty())
+      std::cout << "  (" + boost::algorithm::join(values, "/") + ")";
```
Comment (vtnerd) [inline src/simplewallet/simplewallet.cpp]: No need to create multiple temporary strings, write each portion directly to `std::cout` with `<<`.

File: src/simplewallet/simplewallet.cpp [hunk 2]
```diff
@@ -2154,10 +2160,118 @@ bool simple_wallet::welcome(const std::vector<std::string> &args)
...
+      return values;
+    }
+    std::vector<std::string> operator()(const std::string &v0) const { return {}; }
+    std::vector<std::string> operator()(const std::pair<bool, uint64_t> &v0) const { return {}; }
+  };
+  class input: public boost::static_visitor<bool>
+  {
+    std::string line;
+  public:
+    input(const std::string &s): line(s) {}
```
Comment (vtnerd) [inline src/simplewallet/simplewallet.cpp]: Take by value and move.

File: src/simplewallet/simplewallet.cpp [hunk 3]
```diff
@@ -2154,10 +2160,118 @@ bool simple_wallet::welcome(const std::vector<std::string> &args)
...
+    {
+      std::string prompt = prefix + std::string(e.first ? e.first : "") + tr(" [ENTER to leave it to ") + boost::apply_visitor(to_string(), e.second) + "]";
+      while (1)
+      {
+        std::string line = input_line(prompt, boost::apply_visitor(allowed_values(), e.second));
+        if (std::cin.eof())
+          return false;
+        if (line.empty())
+          break;
+        if (boost::apply_visitor(input(line), e.second))
```
Comment (vtnerd) [inline src/simplewallet/simplewallet.cpp]: `std::move(line)` into `input`.

File: src/wallet/wallet2.cpp [hunk 4]
```diff
@@ -13099,4 +13099,128 @@ uint64_t wallet2::get_bytes_received() const
...
+  decimal_point = boost::get<std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>>(v[0].second).first;
+  cryptonote::set_default_decimal_point(decimal_point);
+  rewrite(m_keys_file, pwd_container.password());
+#endif
+
+  v = {{tr("Minimum output count"), m_min_output_count}, {tr("Minimum output value"), std::make_pair(true, m_min_output_value)}};
+  if (!f(tr("The monero wallet will usually try to consolidate outputs when you send monero. This will tend to concentrate your monero in few larger coins. If you usually need to send several transactions in short order, this can be counter productive as more of your balance will be locked waiting for confirmations. To mitigate this, you can set the wallet to avoid doing so if it would result in fewer than a certain amount of coins of a certain amount. For example, setting the count to 5 and the amount to 2 means that the wallet will try to keep at least 5 coins of 2+ monero."), v)) return;
+  m_min_output_count = boost::get<uint32_t>(v[0].second);
+  m_min_output_value = boost::get<std::pair<bool, uint64_t>>(v[1].second).second;
```
Comment (vtnerd) [inline src/wallet/wallet2.cpp]: Is there any reason why these both are collected in one invocation of the callback function? Couldn't one be grabbed, and then the other? And same thing with the case below? The primary advantage seems to be the display of the one banner message before two of the questions. Couldn't this be done by passing a `nullptr` for the input of the second value so this big banner is still only displayed once? That would drop the need for `std::vector` as input to the callback function.

File: src/wallet/wallet2.cpp [hunk 5]
```diff
@@ -13099,4 +13099,128 @@ uint64_t wallet2::get_bytes_received() const
...
   return m_http_client.get_bytes_received();
 }
+//----------------------------------------------------------------------------------------------------
+void wallet2::configure(const std::function<bool(const char*, std::vector<std::pair<const char*, boost::variant<uint64_t, uint32_t, std::pair<bool, uint64_t>, std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>, bool, std::string>>>&)> &f, const password_container &pwd_container)
+{
+  std::vector<std::pair<const char*, boost::variant<uint64_t, uint32_t, std::pair<bool, uint64_t>, std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>, bool, std::string>>> v;
+
+  v = {{tr("Always prompt before sending a transaction"), m_always_confirm_transfers}};
+  if (!f(NULL, v)) return;
+  m_always_confirm_transfers = boost::get<bool>(v[0].second);
```
Comment (vtnerd) [inline src/wallet/wallet2.cpp]: This design is interesting, but still somewhat awkward. The original value is going into the variant, but is used only for its type, and then you have to copy the value back out again. This funkiness is seen  with `std::pair<bool, std::uint64_t>`, which is just some pseudo-type to differentiate between numerical and money types.

Formally this is actually a `std::function<bool(const std::string&)>` where the result value is "bound" to the function. I.e.:
```c++
v = {
  tr("Always prompt before sending a transaction"),
  std::bind(&parse_bool, _1, std::ref(m_always_confirm_transfers))
};
```

which allows for any `std::string->T` conversion/storage to take place, and separates conversion function from type (currently conversion function is linked to type, thus the `std::pair<bool, std::uint64_t>` stuff).

File: src/wallet/wallet2.cpp [hunk 6]
```diff
@@ -13099,4 +13099,128 @@ uint64_t wallet2::get_bytes_received() const
...
+  m_refresh_type = (RefreshType)boost::get<std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>>(v[0].second).first;
+  rewrite(m_keys_file, pwd_container.password());
+
+  v = {{tr("Refresh from block height"), m_refresh_from_block_height}};
+  if (!f(NULL, v)) return;
+  m_refresh_from_block_height = boost::get<uint64_t>(v[0].second);
+  rewrite(m_keys_file, pwd_container.password());
+
+  v = {{tr("Confirm missing payment ID"), m_confirm_missing_payment_id}};
+  if (!f("If you sometimes have to send monero to a recipient which requires a payment ID, setting this option will prompt if you're sending without one in case you forgot to add one", v)) return;
```
Comment (erciccione) [inline src/wallet/wallet2.cpp]: Could you rephrase this? I'm afraid it could be confusing (to users and translators)

File: src/wallet/wallet2.cpp (hunk 4 above)
Comment (moneromooo-monero) [reply src/wallet/wallet2.cpp]: It's really one setting, composed of two values, so they're meant to be set in relation to each other.

File: src/wallet/wallet2.cpp (hunk 5 above)
Comment (moneromooo-monero) [reply src/wallet/wallet2.cpp]: I was thinking a GUI would probably want to display a widget type which matches the variant type, so a parser function does not really make sense in wallet2. For a bool, it'd have a "selected or not" widget for example.

File: src/wallet/wallet2.cpp (hunk 6 above)
Comment (moneromooo-monero) [reply src/wallet/wallet2.cpp]: How about: setting this option will prompt when you send a payment without a payment ID. This can be useful if you sometimes need to use one and want to make sure you don't forget one when needed.

Comment (moneromooo-monero) [reply src/wallet/wallet2.cpp]: This is not going in for 0.14.1 btw ^_^

Comment (erciccione) [reply src/wallet/wallet2.cpp]: That would be much better IMO.
I guessed wasn't going in for 0.14.1 ;)

File: src/wallet/wallet2.cpp (hunk 5 above)
Comment (moneromooo-monero) [reply src/wallet/wallet2.cpp]: BTW, the value isn't used only for its type. In simplewallet, you can press ENTER and it stays at its current value.

**Comment by moneromooo-monero** [comment]
Actually I'm thinking maybe it should be a single array of these, so that a UI can display them all in a page. With the one-call-per-setting as it is now, it has to be N pages.

File: src/wallet/wallet2.cpp (hunk 5 above)
Comment (vtnerd) [reply src/wallet/wallet2.cpp]: > I was thinking a GUI would probably want to display a widget type which matches the variant type, so a parser function does not really make sense in wallet2. For a bool, it'd have a "selected or not" widget for example.

I'm not sure that the GUI would use as-is because it would force a wizard-style UI instead of a configuration screen (options are thrown to the callback one "group" at a time). Maybe the wizard-style is preferable in this case, I don't know.

If special widgets are needed adding an `enum` alongside the bound function would do it. It would be more readable/auditable than `std::pair<bool, std::uint64_t>`  representing the money type. Which brings me to the next thing - if you bound functions + `enum`s really aren't your thing then considering `struct`s for some of these types so its more self-documenting. And `typedef`ing the large `boost::variant` type would help too.

The other disadvantage to this current approach is that the GUI has to re-define conversion behavior (unless the visitor `struct` is moved to some header). With an `enum` + bound function a UI can just throw any string at the conversion function and check the status `bool`. Unfortunately these functions don't give detailed failure codes, so they may want custom functions for some or all of them anyway.

> BTW, the value isn't used only for its type. In simplewallet, you can press ENTER and it stays at its current value.

This behavior would be unchanged when using bound functions. If the callback is not invoked, then the bound value never changes.

Comment (vtnerd) [reply src/wallet/wallet2.cpp]: Actually, I'm mixed on the `enum` + bound function. Its definitely cleaner in some ways, but the `boost::variant` is an enum + value in one. The bound function seems slightly preferable, but it might be the grossness of `std::pair<bool, std;:uint64_t>` stuck in my head.

Comment (moneromooo-monero) [reply src/wallet/wallet2.cpp]: That sounds maybe like what I'm doing now:

std::vector<std::vector<std::tuple<const char*, boost::variant<uint64_t, uint32_t, std::pair<bool, uint64_t>, std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>, bool, std::string>, std::function<void(std::vector<std::pair<const char*, boost::variant<uint64_t, uint32_t, std::pair<bool, uint64_t>, std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>, bool, std::string>>>&)>>>> wallet2::configure(const password_container &pwd_container)
{
  std::vector<std::vector<std::tuple<const char*, boost::variant<uint64_t, uint32_t, std::pair<bool, uint64_t>, std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>, bool, std::string>, std::function<void(std::vector<std::pair<const char*, boost::variant<uint64_t, uint32_t, std::pair<bool, uint64_t>, std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>, bool, std::string>>>&)>>>> v;

#define SET(type, var) \
  [&](std::vector<std::pair<const char*, boost::variant<uint64_t, uint32_t, std::pair<bool, uint64_t>, std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>, bool, std::string>>> &v0) { \
    var = boost::get<type>(v0[0].second); \
    rewrite(m_keys_file, pwd_container.password()); \
  }

  v.push_back({{tr("Always prompt before sending a transaction"), m_always_confirm_transfers, SET(bool, m_always_confirm_transfers)}});

  v.push_back{{tr("Print ring members before sending a transaction"), m_print_ring_members, SET(bool, m_print_ring_members)}};

Comment (vtnerd) [reply src/wallet/wallet2.cpp]: Only to the `push_back` part, but not the remainder (primarily a switch to enum + bound function or something less crummy than arbitrary `std::pair<bool, std::uint64_t>` types).

Also, having some of these visitors in some header might be useful, if the GUI were truly going to use it at some point.

**Comment by moneromooo-monero** [comment]
In retrospect, it looks like overengineering... I'm not sure whether I should close this.

**Comment by moneromooo-monero** [comment]
Third thought says it's just too complicated, though nifty.
